File: src/1531.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

class Solution {
   public:
    int calc(int x) {
        int retVal = 4;

        if (x == 1) {
            retVal = 1;
        } else if (x < 10) {
            retVal = 2;
        } else if (x < 100) {
            retVal = 3;
        }

        return retVal;
    }
    int getLengthOfOptimalCompression(string s, int k) {
        int retVal = 0;

        int sSize = s.size();

        vector<vector<int>> dp(sSize + 1, vector<int>(k + 1, numeric_limits<int>::max() >> 1));
        dp[0][0] = 0;
        for (int x = 1; x <= sSize; ++x) {
            for (int y = 0; y <= k && y <= x; ++y) {
                if (y > 0) {
                    dp[x][y] = dp[x - 1][y - 1];
                }

                int same = 0;
                int diff = 0;
                for (int z = x; (z >= 1 && diff <= y); --z) {
                    if (s[z - 1] == s[x - 1]) {
                        ++same;
                        dp[x][y] = min(dp[x][y], dp[z - 1][y - diff] + calc(same));
                    } else {
                        ++diff;
                    }
                }
            }
        }
        retVal = dp[sSize][k];

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        string s;
        int k;
    };
    vector<subject> testData{{"aaabcccd", 2}, {"aabbaa", 2}, {"aaaaaaaaaaa", 0}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: s = "aaabcccd", k = 2
     *  Output: 4
     *
     *  Input: s = "aabbaa", k = 2
     *  Output: 2
     *
     *  Input: s = "aaaaaaaaaaa", k = 0
     *  Output: 3
     */

    Solution cSolution;
    int answer = 0;
    for (int x = 0; x < numberOfTestCase; ++x) {
        cout << "Input: s = \"" << testData[x].s << "\", k = " << testData[x].k << "\n";

        answer = cSolution.getLengthOfOptimalCompression(testData[x].s, testData[x].k);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/1531.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def calc(self, x: int) -> int:
        retVal = 0

        if x == 1:
            retVal = 1
        elif x < 10:
            retVal = 2
        elif x < 100:
            retVal = 3
        else:
            retVal = 4

        return retVal

    def getLengthOfOptimalCompression(self, s: str, k: int) -> int:
        retVal = 0

        sSize = len(s)

        dp = [[10**9] * (k + 1) for _ in range(sSize + 1)]
        dp[0][0] = 0

        for x in range(1, sSize + 1):
            for y in range(min(k, x) + 1):
                if y > 0:
                    dp[x][y] = dp[x - 1][y - 1]

                same = 0
                diff = 0
                for z in range(x, 0, -1):
                    if s[z - 1] == s[x - 1]:
                        same += 1
                        value = dp[z - 1][y - diff] + self.calc(same)
                        dp[x][y] = min(dp[x][y], value)
                    else:
                        diff += 1
                        if diff > y:
                            break

        retVal = dp[sSize][k]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for s, k in zip(["aaabcccd", "aabbaa", "aaaaaaaaaaa"], [2, 2, 0]):
            # /* Example
            #  *  Input: s = "aaabcccd", k = 2
            #  *  Output: 4
            #  *
            #  *  Input: s = "aabbaa", k = 2
            #  *  Output: 2
            #  *
            #  *  Input: s = "aaaaaaaaaaa", k = 0
            #  *  Output: 3
            #  */
            logging.info("Input: s = \"%s\", k = %s", s, k)

            retVal = pSolution.getLengthOfOptimalCompression(s, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
